File: iosApp/iosApp/Screens/ApproachesSection.swift
import SwiftUI
import Shared

// Sección APROXIMACIONES de la ficha de escuela — Fase 1 de
// APPROACH_DESIGN.md §6.1. Solo lectura: alta por usuario (grabar, añadir
// chincheta) llega en una fase posterior, sujeta a revisión legal.
//
// El loader vive en SchoolMapSection (no aquí) para que la ficha mini de un
// sector/parking en el mapa también pueda ofrecer "SEGUIR" si existe un
// camino que llega hasta él (ver miniBlockCard).

@MainActor
final class ApproachesLoader: ObservableObject {
    @Published var approaches: [Approach] = []
    @Published var loading = false
    private var loadedSchoolId: String?

    func load(schoolId: String) {
        guard schoolId != loadedSchoolId else { return }
        loadedSchoolId = schoolId
        loading = true
        Task {
            let result = (try? await AppDependencies.shared.container.getApproaches.invoke(schoolId: schoolId)) ?? []
            approaches = result
            loading = false
        }
    }

    /** Fuerza recargar (tras grabar/añadir chincheta), aunque sea la misma escuela. */
    func reload(schoolId: String) {
        loadedSchoolId = nil
        load(schoolId: schoolId)
    }

    /** Primer camino (verificado con preferencia) que llega hasta ese bloque. */
    func approach(to blockId: String) -> Approach? {
        approaches.first { $0.toBlockId == blockId && $0.isVerified }
            ?? approaches.first { $0.toBlockId == blockId }
    }
}

struct ApproachesSection: View {
    @ObservedObject var loader: ApproachesLoader
    let school: School
    let blocks: [Block]
    let isAdmin: Bool
    @Binding var following: Approach?
    @State private var recording = false
    @State private var deleting: Approach?

    private var schoolName: String { school.name }

    var body: some View {
        Group {
            if !loader.approaches.isEmpty || isAdmin {
                VStack(alignment: .leading, spacing: 8) {
                    Text("APROXIMACIONES")
                        .font(Cumbre.mono(10, .bold)).tracking(1.8)
                        .foregroundStyle(Cumbre.ink3)
                        .padding(.horizontal, 16)

                    ForEach(loader.approaches, id: \.id) { a in
                        approachCard(a)
                            .padding(.horizontal, 16)
                    }

                    // Visible SOLO para admin por ahora — la pantalla que abre
                    // es la definitiva, la que verá cualquier usuario cuando
                    // se active para todos (ver APPROACH_DESIGN.md §2.6/§10).
                    if isAdmin {
                        Button {
                            recording = true
                        } label: {
                            Text("+ GRABAR APROXIMACIÓN")
                                .font(Cumbre.mono(11, .bold)).tracking(1.4)
                                .foregroundStyle(Cumbre.terra)
                                .frame(maxWidth: .infinity)
                                .padding(.vertical, 12)
                                .overlay(RoundedRectangle(cornerRadius: 2)
                                    .strokeBorder(Cumbre.terra, style: StrokeStyle(lineWidth: 1, dash: [4, 3])))
                        }
                        .padding(.horizontal, 16)
                    }
                }
                .padding(.vertical, 8)
            }
        }
        .fullScreenCover(isPresented: $recording) {
            ApproachRecordView(school: school, blocks: blocks) {
                loader.reload(schoolId: school.id)
            }
        }
        .alert("¿Borrar «\(deleting?.name ?? "esta aproximación")»?", isPresented: Binding(
            get: { deleting != nil }, set: { if !$0 { deleting = nil } }
        )) {
            Button("CANCELAR", role: .cancel) { deleting = nil }
            Button("BORRAR", role: .destructive) {
                if let a = deleting {
                    Task {
                        _ = try? await AppDependencies.shared.container.approachApi.deleteApproach(approachId: a.id)
                        loader.reload(schoolId: school.id)
                    }
                }
                deleting = nil
            }
        } message: {
            Text("Se borra el camino y todas sus chinchetas. No se puede deshacer.")
        }
    }

    @ViewBuilder
    private func approachCard(_ a: Approach) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            HStack(alignment: .top) {
                Text(a.name ?? "Aproximación")
                    .font(.system(size: 14, weight: .bold)).foregroundStyle(Cumbre.ink)
                Spacer()
                if isAdmin {
                    Button {
                        deleting = a
                    } label: {
                        Image(systemName: "trash").font(.system(size: 13)).foregroundStyle(Cumbre.ink3)
                    }
                }
            }
            Text(summaryLine(a))
                .font(.system(size: 12)).foregroundStyle(Cumbre.ink2)
            HStack {
                Text(a.isVerified ? "✓ VERIFICADA" : "⚠ SIN VERIFICAR")
                    .font(Cumbre.mono(9, .bold)).tracking(1)
                    .foregroundStyle(a.isVerified ? Color(hex: 0x3F6B4A) : Color(hex: 0xB45309))
                Spacer()
                Button {
                    following = a
                } label: {
                    Text("SEGUIR")
                        .font(Cumbre.mono(10, .bold)).tracking(1)
                        .foregroundStyle(.white)
                        .padding(.horizontal, 14).padding(.vertical, 7)
                        .background(Cumbre.terra)
                        .clipShape(RoundedRectangle(cornerRadius: 2))
                }
            }
        }
        .padding(12)
        .background(Cumbre.paper2)
        .overlay(RoundedRectangle(cornerRadius: 2).stroke(Cumbre.rule, lineWidth: 1))
    }

    private func summaryLine(_ a: Approach) -> String {
        var parts: [String] = []
        if let d = a.distanceM?.intValue { parts.append(d >= 1000 ? String(format: "%.1f km", Double(d) / 1000) : "\(d) m") }
        if let asc = a.ascentM?.intValue { parts.append("+\(asc) m") }
        if let dur = a.durationMin?.intValue { parts.append("~\(dur) min") }
        let pinCount = a.pins.count
        if pinCount > 0 { parts.append(pinCount == 1 ? "1 chincheta" : "\(pinCount) chinchetas") }
        return parts.joined(separator: " · ")
    }
}

File: app/src/main/java/com/meteomontana/android/ui/components/ApproachesSection.kt
package com.meteomontana.android.ui.components

import com.meteomontana.android.ui.theme.terraFillColor

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.unit.dp
import com.meteomontana.android.domain.model.Approach
import com.meteomontana.android.ui.theme.EyebrowTextStyle
import com.meteomontana.android.ui.theme.Ok
import com.meteomontana.android.ui.theme.Spacing
import com.meteomontana.android.ui.theme.Terra
import com.meteomontana.android.ui.theme.Warn
import java.util.Locale

/**
 * Sección APROXIMACIONES de la ficha de escuela — Fase 1+ de
 * APPROACH_DESIGN.md §6.1. Espejo de ApproachesSection.swift (iOS).
 *
 * SOLO LECTURA para cualquier usuario; grabar/borrar es admin-only por ahora
 * (§2.6/§10 — pendiente revisión legal de términos antes de abrirlo a todos).
 */
@Composable
fun ApproachesSection(
    approaches: List<Approach>,
    isAdmin: Boolean,
    onFollow: (Approach) -> Unit,
    onRecord: () -> Unit,
    onDelete: (Approach) -> Unit
) {
    if (approaches.isEmpty() && !isAdmin) return

    var deleting by remember { mutableStateOf<Approach?>(null) }

    Column(modifier = Modifier.fillMaxWidth().padding(vertical = Spacing.sm)) {
        Text(
            "APROXIMACIONES",
            style = EyebrowTextStyle,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.padding(horizontal = Spacing.md)
        )
        Spacer(Modifier.padding(top = Spacing.xs))

        approaches.forEach { a ->
            ApproachCard(
                approach = a,
                isAdmin = isAdmin,
                onFollow = { onFollow(a) },
                onDelete = { deleting = a },
                modifier = Modifier.padding(horizontal = Spacing.md, vertical = Spacing.xs)
            )
        }

        // Visible SOLO para admin por ahora — la pantalla que abre es la
        // definitiva, la que verá cualquier usuario cuando se active para
        // todos (ver APPROACH_DESIGN.md §2.6/§10).
        if (isAdmin) {
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = Spacing.md, vertical = Spacing.xs)
                    .clip(RoundedCornerShape(2.dp))
                    .border(1.dp, Terra, RoundedCornerShape(2.dp))
                    .clickable(onClick = onRecord)
                    .padding(vertical = Spacing.sm),
                contentAlignment = Alignment.Center
            ) {
                Text("+ GRABAR APROXIMACIÓN", style = EyebrowTextStyle, color = Terra)
            }
        }
    }

    deleting?.let { a ->
        AlertDialog(
            onDismissRequest = { deleting = null },
            title = { Text("¿Borrar «${a.name ?: "esta aproximación"}»?") },
            text = { Text("Se borra el camino y todas sus chinchetas. No se puede deshacer.") },
            confirmButton = {
                TextButton(onClick = {
                    onDelete(a)
                    deleting = null
                }) { Text("BORRAR", color = MaterialTheme.colorScheme.error) }
            },
            dismissButton = {
                TextButton(onClick = { deleting = null }) { Text("CANCELAR") }
            }
        )
    }
}

@Composable
private fun ApproachCard(
    approach: Approach,
    isAdmin: Boolean,
    onFollow: () -> Unit,
    onDelete: () -> Unit,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .background(MaterialTheme.colorScheme.surface, RoundedCornerShape(2.dp))
            .border(1.dp, MaterialTheme.colorScheme.outline, RoundedCornerShape(2.dp))
            .padding(Spacing.md)
    ) {
        Row(verticalAlignment = Alignment.Top, horizontalArrangement = Arrangement.SpaceBetween,
            modifier = Modifier.fillMaxWidth()) {
            Text(
                approach.name ?: "Aproximación",
                style = MaterialTheme.typography.titleSmall,
                color = MaterialTheme.colorScheme.onSurface,
                modifier = Modifier.weight(1f)
            )
            if (isAdmin) {
                Box(
                    modifier = Modifier.padding(start = Spacing.sm)
                        .clickable(onClick = onDelete),
                    contentAlignment = Alignment.Center
                ) {
                    Text("🗑", color = MaterialTheme.colorScheme.onSurfaceVariant)
                }
            }
        }
        Spacer(Modifier.padding(top = 2.dp))
        Text(
            summaryLine(approach),
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
        Spacer(Modifier.padding(top = Spacing.xs))
        Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.fillMaxWidth()) {
            Text(
                if (approach.isVerified) "✓ VERIFICADA" else "⚠ SIN VERIFICAR",
                style = EyebrowTextStyle,
                color = if (approach.isVerified) Ok else Warn,
                modifier = Modifier.weight(1f)
            )
            Box(
                modifier = Modifier
                    .clip(RoundedCornerShape(2.dp))
                    .background(terraFillColor())
                    .clickable(onClick = onFollow)
                    .padding(horizontal = Spacing.md, vertical = Spacing.xs)
            ) {
                Text("SEGUIR", style = EyebrowTextStyle,
                    color = androidx.compose.ui.graphics.Color.White)
            }
        }
    }
}

private fun summaryLine(a: Approach): String {
    val parts = mutableListOf<String>()
    a.distanceM?.let { d ->
        parts += if (d >= 1000) String.format(Locale.US, "%.1f km", d / 1000.0) else "$d m"
    }
    a.ascentM?.let { parts += "+$it m" }
    a.durationMin?.let { parts += "~$it min" }
    val pinCount = a.pins.size
    if (pinCount > 0) parts += if (pinCount == 1) "1 chincheta" else "$pinCount chinchetas"
    return parts.joinToString(" · ")
}
